src: make --use-largepages a runtime option

gabrielschulhof opened the pull request:
Moves the option that instructs Node.js to-remap its static code to
large pages from a configure-time option to a runtime option. This
should make it easy to assess the performance impact of such a change
without having to custom-build.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

addaleax commented on doc/api/cli.md:
```diff
@@ -863,6 +863,15 @@ environment variables.
 
+### `--use-largepages`
+<!-- YAML
+added: v6.11.0
```
```suggestion
added: REPLACEME
```

gabrielschulhof commented:
@addaleax I added the test and I fixed the doc.
@bnoordhuis I replaced the constant with its value and added the man page entry.

gireeshpunathil commented on src/node.cc:
```diff
@@ -994,6 +984,14 @@ InitializationResult InitializeOncePerProcess(int argc, char** argv) {
+  if (per_process::cli_options->use_largepages) {
+    if (node::IsLargePagesEnabled()) {
+      if (node::MapStaticCodeToLargePages() != 0) {
+        fprintf(stderr, "Reverting to default page size\n");
```
should we confine this printf under debug builds only?

gabrielschulhof replied:
There should be some feedback though if it fails to map, right? Like, there are no performance gains because it failed to map to large pages, not because it succeeded, but the application is such that it doesn't help.

gabrielschulhof replied:
@gireeshpunathil also, how do I confine it to debug builds?

```C++
#ifdef DEBUG
fprintf(...);
#endif  // DEBUG
```
?

gireeshpunathil replied:
I think it is better to place it in the documentation to state that the option is silently ignored if the OS is not configured to support that.

> #ifdef DEBUG

yes.

joyeecheung replied:
I think we generally don't shy away from printing to stderr when a command line option does not work (e.g. warnings are printed when the v8 coverage or profiles cannot be written to disk). For consistency I'd prefer we print to stderr without a DEBUG guard as well.

bnoordhuis replied:
Just to +1 that, I'd rather see this changed back again to the simple boolean option.

gireeshpunathil replied:
in terms of consistency, there are many optimizers / accelerators across jit and gc, that silently pick up the next viable approach when their best bet fails, but do not print out what is happening underneath and what decisions they made. the subject of current PR is an optimization, and imo a runtime should not be verbose around such capabiities, which fits more for a parser or an installer use cases.

On the other hand, I believe the documentation should be the key - following the `madvise`'s example itself, the doc should be explicit in terms of the option as liable to fail because of reasons outside the control of user's and node's scope; and is a directive, not a command. Or else, the abstraction becomes loose with many potential outcomes and side effects.

gabrielschulhof commented:
@bnoordhuis @addaleax @cjihrig @devnexen @gireeshpunathil I changed the implementation to one that accepts three values: 0 (default) to not map, 1 to map but to not report failure, and 2 to map and to report failure on `stderr`.

devnexen commented on src/large_pages/node_large_page.cc:
```diff
@@ -352,7 +352,7 @@ MoveTextRegionToLargePages(const text_region& r) {
   }
 
-  ret = madvise(tmem, size, MADV_HUGEPAGE);
+  ret = madvise(tmem, size, 14 /* MADV_HUGEPAGE */);
```
??

richardlau replied:
@devnexen see https://github.com/nodejs/node/pull/30954#pullrequestreview-332209376

devnexen replied:
Sure but I thought it would be more appropriate to consider the feature unsupported if the system does have the flag.

gabrielschulhof replied:
@devbexen if our releases are built on a system that doesn't have the flag then large pages will be unsupported for all those who download the binary, even if their system has the flag.

devnexen replied:
Good point.

joyeecheung commented on doc/node.1:
```diff
@@ -398,6 +398,13 @@ See
+the target system, this will cause the Node.js static code to be moved onto 2
+MiB pages instead of 4 KiB pages.
+.Pp
+The following values are valid: 0 (do not map), 1 (map, ignore failure, and do not report it), and 2 (map, ignore failure, but report it to standard error).
```
Is there a particular reason that we use numbers here instead of some enum-like strings?

gabrielschulhof replied:
I was thinking that they are faster to compare, so as not to impact startup time. If you don't believe that's a good enough reason, I'm happy to change the values to strings. The precedent for enum-like strings is `--unhandled-rejections`, which accepts `"strict"`, `"warn"`, or `"none"`.

joyeecheung replied:
Compared to all the things that we need to do during startup, the overhead of comparing strings with only 3 possible options seem to be negligible to me, so I think yes, it's better to have more readable options here.

addaleax replied:
I’d also prefer a string here (and this line should be wrapped at 80 characters :))

lundibundi commented on src/node.cc:
```diff
@@ -994,6 +984,16 @@ InitializationResult InitializeOncePerProcess(int argc, char** argv) {
 
+  if (per_process::cli_options->use_largepages == 1 ||
+      per_process::cli_options->use_largepages == 2) {
+    if (node::IsLargePagesEnabled()) {
```
Shouldn't we separately report that largepages are unsupported on this platform?

gabrielschulhof commented:
@addaleax @joyeecheung I updated the possible values to strings: "off", "silent", "verbose".
@lundibundi I applied your nits, and I added reporting for the case where mapping is not supported. I also placed the whole chunk into an ifdef to prevent compiling the code on unsupported platforms. On those platforms and in the verbose case, a message will be written to stderr that they're not supported.

devnexen commented on test/parallel/test-startup-large-pages.js:
```diff
@@ -0,0 +1,30 @@
+                     'invalid value for --use-largepages');
+}
+
+// TODO(gabrielschulhof): Run with --use-largepages=2 and make assertions about
```
You may want to update this.

gabrielschulhof commented:
@devnexen thanks, updated.

lundibundi commented on doc/api/cli.md:
```diff
@@ -863,6 +863,22 @@ environment variables.
+* `off`: No mapping will be attempted. This is the default.
+* `silent`: If supported by the OS, mapping will be attempted. Failure to map
+  will be ignored and will not be reported.
+* `verbose`: If supported by the OS, mapping will be attempted. Failure to map
```
Maybe just `on`? Just to have 'off'/'on' values and avoid possible confusion.

gabrielschulhof replied:
I'll do `off`, `on`, and `verbose`, so we can retain the distinction between reporting and silent operation.

lundibundi replied:
Maybe then have it `off`, `on`, `silent` as I think most of our features with warnings/errors always print them and rarely allow silent mode so it'd make sense to have it verbose by default.

lundibundi commented on src/node_options.cc:
```diff
@@ -748,6 +753,10 @@ PerProcessOptionsParser::PerProcessOptionsParser(
             kAllowedInEnvironment);
 #endif
 #endif
+  AddOption("--use-largepages",
```
This is not behind a flag so wouldn't the `node --use-largepages` be ignored on unsupported platforms?
 I think it will be better to just add `#else return false;` to the `IsLargePagesEnabled` to just have printed that it is unsupported.

gabrielschulhof replied:
Currently we are able to distinguish between large pages not being supported, and not being enabled. To retain that distinction, `IsLargePagesEnabled()` would have to return some enum value.

We have [some ongoing work](https://github.com/intel/iodlr/tree/master/large_page) that provides detailed status indications about the various failures. I'd like to modify that work to include a status `map_unsupported_platform`, and then bring the implementation into Node.js wholesale.

Would it be OK to leave that integration beyond the scope of this PR and land it in a future PR?

lundibundi replied:
Perfectly fine. Great work.

gabrielschulhof commented:
@lundibundi I updated the possible values, leaving "off" as the default, "on" as the verbose option, and "silent" as the silent option.

I left the `#ifdef` in node.cc for now, because I'd like to update our reference implementation such that `IsLargePagesEnabled()` produces a status code indicating that mapping to large pages is not enabled.

gabrielschulhof commented on src/node.cc:
```diff
@@ -994,6 +984,26 @@ InitializationResult InitializeOncePerProcess(int argc, char** argv) {
+    }
+  }
+#else
+  if (per_process::cli_options->use_largepages == "on") {
```
Whoops! This condition has to be `== "on"` and not `!= "silent"` otherwise it complains on all systems where it's not supported 😳

gabrielschulhof commented:
FreeBSD crash is legit.

gabrielschulhof commented:
@devnexen any ideas?

devnexen commented:
will look into it within the week end most likely.

gabrielschulhof commented:
@devnexen Thanks!

gabrielschulhof commented:
@devnexen AFAICT the original implementation also segfaults: https://ci.nodejs.org/job/node-test-commit-freebsd/30323/nodes=freebsd11-x64/consoleFull

Might be because we're using 11.2, which, I'm told on the #freebsd IRC channel, is EOL.

devnexen commented:
sent him a patch, he ought to be able to take it from here.

gabrielschulhof commented:
Landed in 8952105e576315fcb15cd4c9399e950bc191e854. Thanks for the help with FreeBSD, @devnexen!

gabrielschulhof commented:
Dang! I made a copy/paste error while landing. Sorry I omitted you as a reviewer @jasnell!

bcoe commented:
@gabrielschulhof we're seeing a bit of a hiccup in our nightly linux tests:

```bash
20:01:53 /bin/sh: 1: realpath: not found
20:01:53 gyp: Call to 'realpath src/large_pages/ld.implicit.script' returned exit status 127 while in /home/iojs/build/workspace/node-test-commit-linux-coverage-daily/nodes/benchmark/node.gyp.
```

Which seems to crop up somewhere between December 21st and December 22nd when this landed. I'm guessing there might be a workaround in changing our compiler flags? Wondering if this is expected however (would guess moving this to a flag was meant to be a noop?).
